### ISSUE TITLE
VS 17.3 Preview 3 Regression from Preview 2 with ref-returning from a struct

### ISSUE THREAD

**Issue by fitdev** [dotnet/roslyn#62901]
**Version Used**: 
VS 2022 17.3 Preview 3. Worked without errors in VS 2022 17.3 Preview 2 and earlier

**Steps to Reproduce**:
Consider [this code](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdAEoCuMCUYAplQMID2YADlADYsATgGVhANygBjFigDcRUhWr1GzNgElGwzjzFDJM+UUUBmMkJYQAJpxj8AnmQYoIAMxZkUCIXSkIyMQBzVkYyAG8TQjIYinMpO28yWACRKAAvFgB5NzIAXi8Mlk43AApgBwQWAEoyAGpCzJLSlAALTiEEWoaUIua2jq6FQiJYuIsrW3snCqqyLWsWAA9hscxzSxs7Ry92zrIAMSFuVdj1ia3p3cGyABVOU5iiAF8ogG0RHz8EABkIB04dAQpT+AKBAGlYNYqABRJY8fjSKBdAC6ZmcMFcHi8X38gRYIRYjBhMDorCEEAQHQiUQSmICKTIABEWG4IHR+AgALIQJbBUIILgMAIFbAkYajM7mflEhAksnCSkdUqcYAAKxYeJQ3CsQiCtUi0TGZHYdCElkYC2W+TI4slYzcUCWLGsZHKlRYACoyDwfDaAEJ0NweIQG+3G8QQIReAkoG1clhgDoOHlCNoQfgcTZVEQ8CAwUqWXIAVUx7jYAEEUDRWQAeGWMAB8pV9obgzNZ7M5PL5BIFQsY1UexoA9CODgJ+GQEK1PMAgyGyPnXSgWAEBwFZ5YyMBUC6yHYD+rNQFtawo0Fw2MNza7UbjYcoGmEA2RWRFmyOQhhzFXveyH+4ZvBOLD8NYORuKuwIkNUaL/oyprmrKVorEBIFgRBUEtH0ZQpNUsHhghgKMMOwFQKB4HBlhvRNLhg5kN62AEf+5yOs6rqzJ4gbBsIbwsp+3a8q+N7eq+VBpJkOQoqR6GUZBa7YbRpR4QxZBMXBazSn2sqPs+r4SixGysvihJhIhFoio2Ey5AmSZCCmUbppm7DZiwub5oWxkTnp2mMO2/FdtyQm+YKxFdG85nITAixLGQAC0ZC4NJZBjmQtZxQlwjHNG7AiAAHKYAAsADsIAmvmMCcAEdCrkuEwoF+B65AA5LZyapk5WZWDmeYwPWIXNkWJkCu2eHNckMDTq0UBxnSVRLAEwCauytXImQkBOMsPCcLVkZCFAEDAIIcZFsIRIyBxTh5hSrBVNGzWnRaMjjYCCC9IsTVTSwT7vpq/BRpSUCHigCQ8Cw4bhucmCFcyM3baupQGgBkPmMAnCcFOXKcOILAAHLLMCtR5FZkWWtFyx1A0tYmmFw5QzDNZYUTJNmhZKG3gZzxAA==):

```cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

public readonly unsafe struct Segment {

    public const int SizeOf = sizeof(byte) + sizeof(short) + sizeof(short);

    public readonly byte Index;
    public readonly short From;
    public readonly short To;
    
}

[StructLayout(LayoutKind.Explicit)]
public unsafe struct SegmentEnumerator {

const int DefaultMaxSegmentCount = 20;

    public SegmentEnumerator(object somearg) {
      CurrentIndex = 0;
      fixed (byte* ptr = Buffer) {
        var segs = MemoryMarshal.CreateSpan(ref Unsafe.AsRef<Segment>(ptr), DefaultMaxSegmentCount);
        //Fill the buffer and set Count here based on object somearg
        Count = 0;
        FirstSegment = default;
      }
    }

    [FieldOffset(0)]
    int CurrentIndex;
    [FieldOffset(sizeof(int))]
    int Count;
    [FieldOffset(sizeof(int) * 2)]
    public fixed byte Buffer[DefaultMaxSegmentCount * Segment.SizeOf];
    [FieldOffset(sizeof(int) * 2)]
    public Segment FirstSegment;

    public ref Segment Current => ref MemoryMarshal.CreateSpan(ref FirstSegment, DefaultMaxSegmentCount)[CurrentIndex - 1]; // <-- error CS8347: Cannot use a result of 'MemoryMarshal.CreateSpan<Segment>(ref Segment, int)' in this context because it may expose variables referenced by parameter 'reference' outside of their declaration scope
    
    public void Dispose() { }
    public bool MoveNext() => CurrentIndex++ < Count;
    public void Reset() => CurrentIndex = 0;

}
```

**Expected Behavior**: Code compiles as in VS 2022 17.3 Preview 2 or earlier.

**Actual Behavior**: Compile fails in VS 2022 17.3 Preview 3: CS8347: Cannot use a result of 'MemoryMarshal.CreateSpan<Segment>(ref Segment, int)' in this context because it may expose variables referenced by parameter 'reference' outside of their declaration scope

**Comment by Sergio0694** [comment]
The fact it fails to build on sharplab should be expected behavior, as the BCL version it uses is not the same as .NET 7, so it is likely using the old `MemoryMarshal.CreateSpan` signature, which doesn't take a `scoped ref`. In that case, that compile error is the correct result for that line of code. This should build fine when using an updated .NET 7 SDK with the proper annotation for that API. On the same note: you should most definitely not rely on that anyway, but rather expose proper tracking for this API so that it's not dangerous for consumers. That is, you should annotate that `Current` property with the `[UnscopedRef]` attribute (see https://github.com/dotnet/runtime/issues/72074). That will both solve the build error in your configuration, but also enforce lifetime tracking for callers.

Either way this doesn't seem like a compiler bug to me 🙂

**Comment by jcouv** [comment]
@cston Was this intentional?

**Comment by jaredpar** [comment]
Closing as this behavior is "By Design". 

The issue here is that the `net6.0` version of `MemoryMarshal.CreateSpan` has no `scoped` annotation. That means it appears as `Span<T> CreateSpan<T>(ref int length)`. According to the updated rules the parameter can be captured into the return. The new `net7.0` version of the API has the correct annotations to make this work. 

For scenarios like this where you want to cross compile with new language version and old TF then you'll need to use `MemoryMarshal.AsReference` to work around this change. 

```csharp
ref MemoryMarshal.CreateSpan(
  ref MemoryMarshal.AsReference(ref FirstSegment) 
  DefaultMaxSegmentCount)[CurrentIndex - 1];
```

**Comment by JeremyMorton** [comment]
> For scenarios like this where you want to cross compile with new language version and old TF then you'll need to use `MemoryMarshal.AsReference` to work around this change.
> 
> ```cs
> ref MemoryMarshal.CreateSpan(
>   ref MemoryMarshal.AsReference(ref FirstSegment) 
>   DefaultMaxSegmentCount)[CurrentIndex - 1];
> ```

I want to try this workaround, but I am not sure what you are referring to with `MemoryMarshal.AsReference`; I can't find any reference to it. `MemoryMarshal.AsRef` takes a span, which is what I'm trying to create in the first place. :)

**Comment by jaredpar** [comment]
> I want to try this workaround, but I am not sure what you are referring to with MemoryMarshal.AsReference; I can't find any reference to it. MemoryMarshal.AsRef takes a span, which is what I'm trying to create in the first place. :)

This is because I should have written `Unsafe.AsRef`. 😦 

Overall though my statement is inaccurate. After we saw more examples of this in the wild we decided to directly address the problem. When calling APIs that are compiled with the pre-C# 11 version of the rules we interpret them using the old rules. So in this case we'll do the right thing and the code can exist as it did in C# 10 and before.
